archunit/src/test/java/com/tngtech/archunit/core/importer/ClassFileImporterAccessesTest.java

```diff
@@ -754,6 +776,247 @@ public void classes_know_method_calls_to_themselves() {
+                        SourceCodeLocation.of(javaClass, 7),
+                        SourceCodeLocation.of(javaClass, 13));
+    }
+
+    @Test
+    public void imports_try_catch_blocks_from_same_starting_label() {
+        assumeTrue(!System.getProperty("java.version").startsWith("1.7."));
+
+        // this try-catch-combination creates a try-catch and a try-finally block from the same starting label
+        // we test that the second recorded block is not cleared out by accident when the first block is closed
+        @SuppressWarnings({"unused", "ConstantConditions", "TryFinallyCanBeTryWithResources", "UnnecessaryReturnStatement"})
+        class SomeClass {
+            private void method(final int first, final boolean second) {
+                try {
+                    Socket client = new Socket("", 0);
+                    BufferedReader reader = new BufferedReader(null);
+                    try {
+                        return;
+                    } finally {
+                        reader.close();
+                    }
+                } catch (Exception ignored) {
+                }
+            }
+        }
+
+        JavaMethod method = new ClassFileImporter().importClass(SomeClass.class).getMethod("method", int.class, boolean.class);
+
+        assertThat(method.getTryCatchBlocks()).hasSize(3);
+    }
+
+    @Test
+    public void imports_try_catch_block_with_multiple_caught_throwables() {
+        @SuppressWarnings("unused")
+        class SomeClass {
+            void method() {
+                try {
+                    new Object();
+                } catch (IllegalStateException | IllegalArgumentException ignored) {
+                } catch (UnsupportedOperationException ignored) {
+                    System.out.println("unsupported");
+                } finally {
+                    System.out.println("finally");
+                }
+            }
+        }
+
+        JavaMethod method = new ClassFileImporter().importClass(SomeClass.class).getMethod("method");
+        TryCatchBlock tryCatchBlock = getOnlyElement(method.getTryCatchBlocks());
+
+        assertThatTypes(tryCatchBlock.getCaughtThrowables()).matchInAnyOrder(
+                IllegalStateException.class,
+                IllegalArgumentException.class,
+                UnsupportedOperationException.class
+        );
+    }
+
+    @Test
+    public void imports_complex_nested_try_catch_blocks() {
+        JavaClass javaClass = new ClassFileImporter().importClass(ClassWithComplexTryCatchBlocks.class);
+
+        JavaConstructor constructor = javaClass.getConstructor();
+        Set<TryCatchBlock> tryCatchBlocks = constructor.getTryCatchBlocks();
+
+        assertThat(tryCatchBlocks)
+                .hasSize(3)
+                .areExactly(1, tryCatchBlock()
+                        .declaredIn(constructor)
+                        .catching(Exception.class)
+                        .atLocation(ClassWithComplexTryCatchBlocks.class, 14)
```

Comment:
It might be interesting that the outer `TryCatchBlock` 's `accessesContainedInTryBlock` contains the inner one's.
(But it's admittedly tedious to describe in the test.)

Reply:
Never mind, I later saw that this is implicitly covered in the `@Test` `all_accesses_know_which_exceptions_are_handled`.